File: Documentation/4.0/Samples/csharp/Raven.Documentation.Samples/ClientApi/Session/Configuration/OptimisticConcurrency.cs
using Raven.Client.Documents;
using Raven.Client.Documents.Session;
using Raven.Documentation.Samples.Orders;

namespace Raven.Documentation.Samples.ClientApi.Session.Configuration
{
    public class OptimisticConcurrency
    {
        public OptimisticConcurrency()
        {
            using (var store = new DocumentStore())
            {
                #region optimistic_concurrency_1
                using (IDocumentSession session = store.OpenSession())
                {
                    // Enable optimistic concurrency for this session
                    session.Advanced.UseOptimisticConcurrency = true;

                    // Save a document in this session
                    Product product = new Product { Name = "Some Name" };
                    session.Store(product, "products/999");
                    session.SaveChanges();

                    // Modify the document 'externally' by another session 
                    using (IDocumentSession otherSession = store.OpenSession())
                    {
                        Product otherProduct = otherSession.Load<Product>("products/999");
                        otherProduct.Name = "Other Name";
                        otherSession.SaveChanges();
                    }

                    // Trying to modify the document without reloading it first will throw
                    product.Name = "Better Name";
                    session.SaveChanges(); // This will throw a ConcurrencyException
                }
                #endregion

                #region optimistic_concurrency_2
                // Enable for all sessions that will be opened within this document store
                store.Conventions.UseOptimisticConcurrency = true;

                using (IDocumentSession session = store.OpenSession())
                {
                    bool isSessionUsingOptimisticConcurrency = session.Advanced.UseOptimisticConcurrency; // will return true
                }
                #endregion

                #region optimistic_concurrency_3
                using (IDocumentSession session = store.OpenSession())
                {
                    // Store document 'products/999'
                    session.Store(new Product { Name = "Some Name" }, id: "products/999");
                    session.SaveChanges();
                }

                using (IDocumentSession session = store.OpenSession())
                {
                    // Enable optimistic concurrency for the session
                    session.Advanced.UseOptimisticConcurrency = true;

                    // Store the same document
                    // Pass 'null' as the changeVector to turn OFF optimistic concurrency for this document
                    session.Store(new Product { Name = "Some Other Name" }, changeVector: null, id: "products/999");
                    
                    // This will NOT throw a ConcurrencyException, and the document will be saved
                    session.SaveChanges();
                }
                #endregion

                #region optimistic_concurrency_4
                using (IDocumentSession session = store.OpenSession())
                {
                    // Store document 'products/999'
                    session.Store(new Product { Name = "Some Name" }, id: "products/999");
                    session.SaveChanges();
                }

                using (IDocumentSession session = store.OpenSession())
                {
                    // Disable optimistic concurrency for the session 
                    session.Advanced.UseOptimisticConcurrency = false; // This is also the default value

                    // Store the same document
                    // Pass 'string.Empty' as the changeVector to turn ON optimistic concurrency for this document
                    session.Store(new Product { Name = "Some Other Name" }, changeVector: string.Empty, id: "products/999");
                    
                    // This will throw a ConcurrencyException, and the document will NOT be saved
                    session.SaveChanges();
                }
                #endregion
            }
        }
    }
}

File: Documentation/4.0/Samples/nodejs/client-api/session/configuration/optimisticConcurrency.js
import { DocumentStore, DocumentConventions } from "ravendb";

const store = new DocumentStore();
class Product { }

async function example() {
    
    //region optimistic_concurrency_1
    // Enable optimistic concurrency for this session
    const session = store.openSession();
    session.advanced.useOptimisticConcurrency = true;
    
    const product = new Product();
    product.name = "Some Name";

    // Save a document in this session
    await session.store(product, "products/999");
    await session.saveChanges();

    {
        // Modify the document 'externally' by another session 
        const anotherSession = store.openSession();
        
        const otherProduct = await anotherSession.load("products/999");
        otherProduct.name = "Other Name";
        await anotherSession.saveChanges();
    }

    // Trying to modify the document without reloading it first will throw
    product.name = "Better Name";
    await session.saveChanges(); // This will throw a ConcurrencyException
    //endregion
                
    //region optimistic_concurrency_2
    // Enable for all sessions that will be opened within this document store
    store.conventions.useOptimisticConcurrency = true;

    {
        const session = store.openSession();
        const isSessionUsingOptimisticConcurrency
            = session.advanced.useOptimisticConcurrency; // true
    }
    //endregion

    //region optimistic_concurrency_3
    {        
        const session = store.openSession();
        
        const product = new Product();
        product.name = "Some Name";

        // Store document 'products/999'
        await session.store(product, "products/999");
        await session.saveChanges();
    }
    {
        const session = store.openSession();
        
        // Enable optimistic concurrency for the session
        session.advanced.useOptimisticConcurrency = true;

        const product = new Product();
        product.name = "Some Other Name";

        // Store the same document
        // Pass 'null' as the changeVector to turn OFF optimistic concurrency for this document
        await session.store(product, "products/999", { "changeVector": null });

        // This will NOT throw a ConcurrencyException, and the document will be saved
        await session.saveChanges();
    }
    //endregion

    //region optimistic_concurrency_4
    {
        const session = store.openSession();
        
        const product = new Product();
        product.name = "Some Name";

        // Store document 'products/999'
        await session.store(product, "products/999");
        await session.saveChanges();
    }
    {
        const session = store.openSession();
        
        // Disable optimistic concurrency for the session 
        session.advanced.useOptimisticConcurrency = false; // This is also the default value

        const product = new Product();
        product.name = "Some Other Name";

        // Store the same document
        // Pass an empty string as the changeVector to turn ON optimistic concurrency for this document
        await session.store(product, "products/999", { "changeVector": "" });

        // This will throw a ConcurrencyException, and the document will NOT be saved
        await session.saveChanges();
    }
    //endregion
}
